packages/firestore/src/model/path.ts

```diff
@@ -63,31 +46,34 @@ abstract class Path {
   }
 
   /**
-   * Constructs a new instance of Path using the same concrete type as `this`.
-   * We need this instead of using the normal constructor, because polymorphic
-   * `this` doesn't work on static methods.
+   * Abstract constructor method to construct an instance of B with the given
+   * parameters.
    */
-  private construct(
+  protected abstract construct(
     segments: string[],
     offset?: number,
     length?: number
-  ): this {
-    const path: this = Object.create(Object.getPrototypeOf(this));
-    path.init(segments, offset, length);
-    return path;
-  }
+  ): B;
+
+  /**
+   * Returns a String representation.
+   *
+   * Implementing classes are required to provide deterministic implementations as
+   * the String representation is used to obtain canonical Query IDs.
+   */
+  abstract toString(): string;
 
   get length(): number {
     return this.len;
   }
 
-  isEqual(other: Path): boolean {
-    return Path.comparator(this, other) === 0;
+  isEqual(other: this): boolean {
+    return BasePath.comparator(this, other) === 0;
   }
 
-  child(nameOrPath: string | this): this {
+  child(nameOrPath: string | this): B {
     const segments = this.segments.slice(this.offset, this.limit());
-    if (nameOrPath instanceof Path) {
+    if (nameOrPath instanceof BasePath) {
```

Comment:
Maybe make this `nameOrPath instanceof this.constructor`. Or leave as is, if you don't like regressions :)

Reply:
Neat trick, but that breaks TypeScript's type narrowing (so `nameOrPath.foreach()` doesn't work since TypeScript still thinks it may be a string.

But I was able to change nameOrPath to be `string | B` instead of `string | this` which gives us better compile-time type-safety.